### PR TITLE
Dataloader now shuffles the shards and documents within

### PR DESCRIPTION
**Pull request by fraserlove**
Added a function to separate a shard into documents, shuffle the documents, and then concatenate the documents back to a single shard. Individual shards are also shuffled. This removes the unusual periodicity present within the fineweb-edu dataset and results in a smoother training loss trajectory (see image below).

### CODE REVIEW

File: train_gpt2.py [hunk 1]
```diff
@@ -231,10 +232,22 @@ def __init__(self, B, T, process_rank, num_processes, split):
...
+            np.random.shuffle(documents)
+            shard = torch.cat(documents) # concatenate the documents back together
+        return shard
+
     def reset(self):
         # state, init at shard zero
         self.current_shard = 0
-        self.tokens = load_tokens(self.shards[self.current_shard])
+        if self.split == "train":
+            np.random.shuffle(self.shards)
```
Comment (lukasugar) [inline train_gpt2.py]: I suggest you create a random generator, initialized differently on each rank, so the results can be more reproducible.
Can use: https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.shuffle.html#numpy.random.Generator.shuffle
E.g.:
```
rng = np.random.default_rng(self.process_rank + 1337)
...
rng.shuffle(self.shards)
```

Comment (fraserlove) [reply train_gpt2.py]: Thanks for your suggestion.

From the way I understand it, shouldn't the random generator just be
```
rng = np.random.default_rng(1337)
```
as the shuffled data should be the same across all processes? Then each process takes a different batch from this shuffled data using the process_rank as an offset.

Comment (lukasugar) [reply train_gpt2.py]: You're right, I've made a mistake!

**Comment by dustinwloring1988** [comment]
@lukasugar I have been training multiply model and have event customized the model. I was seeing if you would be interest in working together on a new model. I have a FastTokenizer with 128K vocab and new special tokens. If you are interested I can share more information in a PM
